Feature/retailer/add product

Kwang-Kian-Hui opened the pull request.

Kwang-Kian-Hui committed:
retailer product list feature

class object and dto file only.

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/product-list

Kwang-Kian-Hui committed:
structure for implementing product list feature

necessary files (without actual code) and structure for retrieving product list data from firestore.

Kwang-Kian-Hui committed:
widgets for implementing product list feature

retailer drawer widget.
basic structure of retailer home page, product list view and product list item.

Kwang-Kian-Hui committed:
implement product list feature

1) core/shared/provider - included providers for provider
2) product remote service - query nested firebase collection data
3) retailer home page - initState -> get userID(uid) and pass uid to getProductList

Kwang-Kian-Hui committed:
file movement - firestore_failures.dart

Moved firestore_failures.dart and the freezed file to core/infrastructure since these failures can be used for both retailer and consumer query failures. Updated file path for necessary files

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/product-list

Kwang-Kian-Hui committed:
implement add product notifier and state

Added add_product_notifier and add_product_state.
Edited product repository and remote service to implement add product feature.
Added 'product' and 'retailer' folder in core/application and core/infrastructure to group files.

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
changed import file path

Kwang-Kian-Hui committed:
added firestore failures

added out of resource and unavailable service exceptions

Kwang-Kian-Hui committed:
Implement retrieve product list logic and form page

1)Added retrieve product list logic and display on retailer home page (subject to further changes, mainly used for visualizing and testing here)
2)Added product form page, error messages and text fields with formatting
3)Changes to FormTextField
4)Added extended_masked_text package (for usual price and discount price)

Changes to FormTextField:
1)Added controller field
3)Added prefix text field

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
Resolve conflict after main branch merge

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
Merge duplicate product and product dto from main

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
Implement add product logic

Add product logic implemented, but error with retrieving product list due to firestore timestamp type needing to convert to FieldValue, will be fixed in next commit

Kwang-Kian-Hui committed:
Implement display product logic

Retrieved product from firestore and display in retailer home page.

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
Implement bug fixes and removed unused imports

sinhaugoh commented on lib/core/infrastructure/firestore_failures.dart:
```diff
@@ -7,6 +7,8 @@ class FirestoreFailures with _$FirestoreFailures {
+  const factory FirestoreFailures.serviceUnavailable() = ServiceUnavailable;
   const factory FirestoreFailures.unknown() = Unknown;
   const factory FirestoreFailures.noConnection() = NoConnection;
 }
```
too much failure i think.

sinhaugoh commented on lib/core/presentation/routes/app_router.dart:
```diff
@@ -18,7 +19,9 @@ import 'package:deall/splash/splash_page.dart';
+    MaterialRoute(page: AddProductPage),
     MaterialRoute(page: RetailerSignUpPage),
+    MaterialRoute(page: AddProductPage),
     MaterialRoute(page: RetailerProfilePage),
```
duplicate add product page

sinhaugoh commented on lib/core/shared/providers.dart:
```diff
@@ -19,6 +25,17 @@ final imagePickingRemoteServiceProvider =
 
+final productListNotifierProvider =
+    StateNotifierProvider<ProductListNotifier, ProductListState>(
+        (ref) => ProductListNotifier(ref.watch(productListRepoProvider)));
```
The product list shouldn't be in core, coz retailer use stream of product, but consumer use future

sinhaugoh commented on lib/retailer/application/add_product_form_notifier.dart:
```diff
@@ -0,0 +1,128 @@
+        name: state.name,
+        usualPrice: state.usualPrice,
+        discountedPrice: state.discountedPrice,
+        image: state.imageFile.toString(),
```
this shouldn't be working need to use uploadProductImage in imagePicking repo, refer to retailer sign up notifier.

Kwang-Kian-Hui committed:
Restructure add product logic

Improved image picking and restructure codes. Removed duplicate add product route

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
resolve conflicts and file path changes

Merged main branch to this branch. Resolve path and other conflicts

Kwang-Kian-Hui committed:
Implement validator and other changes

Added ValueFailure.invalidPriceValue() for usual and discounted price.
Removed some firestore failures.
fixed bug where products with no image does not get added.

Kwang-Kian-Hui committed:
Changed provider file based on feedback

Moved product list provider to retailer/shared/provider.dart

Kwang-Kian-Hui commented:
Changes made, tested and works exactly as it should for products with and without image

sinhaugoh commented on lib/core/infrastructure/product/product_dto.dart:
```diff
@@ -65,10 +65,22 @@ class ProductDTO with _$ProductDTO {
+      'availability': availability,
+      'dateModified': dateModified,
+    };
+  }
```
dont need to do tojson and fromJson manually, can just remove

sinhaugoh commented on lib/core/application/value_validator.dart:
```diff
@@ -34,3 +34,11 @@ Either<ValueFailure, Unit> validatePostal(String input) {
+
+Either<ValueFailure, Unit> validateDiscountedPrice(double input, double usualPrice) {
+  return input >= 0.0 && input <= 10000 && input < usualPrice ? right(unit) : left(const ValueFailure.invalidPriceValue());
+}
```
for validation of usual price and discounted price u can check if it is empty first, if it is then return empty failure, then proceed to check with invalid value

sinhaugoh commented on lib/retailer/application/add_product_form_notifier.dart:
```diff
@@ -52,44 +57,91 @@ class AddProductFormNotifier extends StateNotifier<AddProductFormState> {
-      final newProduct = Product(
-        id: '',
+      final uid = _authRepository.getUserId();
+      final pid = _productRepository.getProductId(uid);
```
try to use productId instead. make the variable name meaningful

sinhaugoh commented on lib/retailer/application/add_product_form_notifier.dart:
```diff
@@ -52,44 +57,91 @@ class AddProductFormNotifier extends StateNotifier<AddProductFormState> {
+          );
+        });
+      }
+
```
can just use newProduct.copyWith(image: filePath)

sinhaugoh commented on lib/retailer/infrastructure/product_repository.dart:
```diff
@@ -25,24 +26,15 @@ class ProductListRepository {
     }
   }
 
+  String getProductId(String uid){
```
can rename to generateNewProductId? sounds more meaningful

sinhaugoh reviewed (changes requested):
overall looks good, just minor change needed

Kwang-Kian-Hui committed:
Changes made following requested changes in review

Kwang-Kian-Hui committed:
Merge branch 'main' into feature/retailer/add-product

Kwang-Kian-Hui committed:
removed unused codes and other changes

add ignore statements and remove unused imports and code line

sinhaugoh commented on lib/retailer/application/add_product_form_notifier.dart:
```diff
@@ -72,6 +82,18 @@ class AddProductFormNotifier extends StateNotifier<AddProductFormState> {
+        orElse: () {},
+      ),
+      (r) => stateCopy = stateCopy.copyWith(discountedPriceErrorMessage: null),
+    );
```
need to redo this part, logic error

Kwang-Kian-Hui committed:
Changed validation check for usual and discounted price

Kwang-Kian-Hui committed:
Fixed validation check

sinhaugoh reviewed (approved):
everything correct
